File: module/channel.py
#!/bin/env python3
# SoftTSC - Software MPT1327 Trunking System Controller
# Copyright (C) 2013-2014 Paul Banks (http://paulbanks.org)
# 
# This file is part of SoftTSC
#
# SoftTSC is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# SoftTSC is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with SoftTSC.  If not, see <http://www.gnu.org/licenses/>.
#

"""Implementation of MPT1327 channel controller"""

import sys
import logging
from libmpt1327modem import MPT1327Modem
import mpt1327 as mpt
from queue import Queue

class TRAF:
  """Traffic SYNC codeword"""
  def cw(self):
    return 1 # Special meaning
  def __str__(self):
    return "TRAFFIC"

class TXItem:
  """TX queue item"""
  def __init__(self, cw, txcompletionitem=None, rxcompletionitem=None):
    self.cw = cw
    self.txcomplete = txcompletionitem
    self.rxcomplete = rxcompletionitem

class TXCompletionItem:
  """TX completion item"""
  def __init__(self, cfunc, data=None):
    self.cfunc = cfunc
    self.data = data
  def complete(self, ch):
    return self.cfunc(self.data, ch)

class RXCompletionItem:
  """Solicited RX completion item"""
  def __init__(self, size, cfunc, data=None):
    self.size = size
    self.remaining = size
    self.cfunc = cfunc
    self.data = data
    self.cws = []
    self.slottimeout = 4 # 64-bit slots

  def addcw(self, ch, cw):
    self.cws += [cw]
    self.remaining -= 1
    self.slottimeout = 4 # Reset timeout
    if self.remaining>0:
      return 0 # More codewords needed
    else:
      self.cfunc(self.data, 0, ch, self.cws)
      return 1

  def tick(self, ch):
    self.slottimeout -= 1
    if self.slottimeout==0:
      ch.rxcomplete = None # We timed out of our slot so no point waiting
      self.cfunc(self.data, 1, ch, None)      

class Channel:
  """MPT1327 channel controller"""

  def __init__(self, syscode, channelnumber, rxfunc, rxfuncdata, txstate=0):
    self.syscode = syscode
    self.channelnumber = channelnumber
    self.rxfunc = rxfunc
    self.rxfuncdata = rxfuncdata
    self.ahlcount = 0           # Aloha frame counter
    self.rxcomplete = None      # RX completion object for solicited message
    self.txcomplete = None      # TX completion object
    self.txqueue = Queue()      # Control channel transmission queue
    self.txtrafqueue = Queue()  # Traffic channel transmission queue
    self.txstate = txstate      # TX state machine state
    self.txreserved = 0         # Slot reservations
    self.txreserveditem = None  # RX completion object holder
    channelId = "TSC-Ch.%d" % channelnumber
    self.modem = MPT1327Modem(channelId, Channel._rxcv, Channel._txcv, self)
    self.logger = logging.getLogger(__name__)
    self.morse = None
    self.lastcw = None

  def _tick(self):
    if self.rxcomplete:
      self.rxcomplete.tick(self)

  def _txcvimpl(self):
    cw = None
    n = 0

    # Log last transmitted
    if self.lastcw:
      self.logger.debug("TX[%d]: %s", self.channelnumber, self.lastcw)
      self.lastcw = None

    # Transmission provides our ticker for time-outs
    self._tick()

    # Call transmission complete item
    if self.txcomplete:
      nextstate = self.txcomplete.complete(self)
      self.txcomplete = None
      if nextstate is not None:
        self.txstate = nextstate

    # Queue received item for action by receiver
    if self.txreserveditem:
      self.rxcomplete = self.txreserveditem
      self.txreserveditem = None

    if self.txstate==0: # STATE:0 - DECIDE OR BEGIN CODEWORD

      # Morse ident transmission - begins outside MPT1327 frame only
      if not self.morse==None and self.ahlcount==0:
        self.modem.morse(self.morse, Channel._morse_done, self)
        self.morse = None
        self.txstate = 4
      else:
        cw = mpt.CCSC(self.syscode)
        self.txstate = 1

    elif self.txstate==1: # STATE:1 - CONTROL CHANNEL CODEWORD
      self.txstate = 0
      
      # Frame handling
      if self.ahlcount==0:
        self.ahlcount = 5
        n = self.ahlcount
      self.ahlcount -= 1
      if self.ahlcount<0:
        self.ahlcount = 0

      # Withdrawn slots handling (7.2.5)
      if self.txreserved:
        self.txreserved -= 1
      if self.txreserved: # Multi slot...
          cw = mpt.AHY(0, mpt.DUMMYI, mpt.DUMMYI, 0, 0, 0, 0 ,0)

      # Queue handling
      elif not self.txqueue.empty():
        o = self.txqueue.get_nowait()
        cw = o.cw
        self.txcomplete= o.txcomplete
        self.txreserved = 0
        if o.rxcomplete:
          self.txreserved = o.rxcomplete.size
          self.txreserveditem = o.rxcomplete
      else:
        cw = mpt.ALH(0,0,self.channelnumber,6,0,0,n)

    elif self.txstate==2: # STATE:2 - TRAFFIC CHANNEL QUIESCENT STATE
      self.alhcount = 0
      if not self.txtrafqueue.empty(): 
        #cw = mpt.DCSC(self.syscode)
        cw = TRAF()
        self.txstate = 3

    elif self.txstate==3: # STATE:3 - TRAFFIC CHANNEL CODEWORD STATE
      self.txstate = 2
      o = self.txtrafqueue.get_nowait()
      cw = o.cw
      self.txcomplete = o.txcomplete
      self.txreserved = 0
      if o.rxcomplete:
        self.txreserved = o.rxcomplete.size
        self.txreserveditem = o.rxcomplete

    elif self.txstate==4: # STATE:4 - MORSE IDENT/ETC (TODO)
      pass

    # Encode codeword and return
    if cw:
      self.lastcw = cw
      return cw.cw()
    else:
      return 0

  def _rxcvimpl(self, cw):

    # If we're expecting a response in reserved slot(s)...
    if self.rxcomplete:
      self.logger.debug("RX[%d] RSVD: 0x%x", self.channelnumber, cw)
      if self.rxcomplete.addcw(self, cw):
        self.rxcomplete = None
      return 0

    # Otherwise this is a random access which *must* be an address codeword
    else:
      o = mpt.RUtoTSCDecode(cw)
      self.logger.debug("RX[%d]: 0x%x %s", self.channelnumber, cw, o)
      if (o):
        self.rxfunc(self.rxfuncdata, self, o)

  def _morse_done(self):
    self.txstate = 0

  def _txcv(self):
    try:
      return self._txcvimpl()
    except:
      self.logger.exception("TX[%d] Exception", self.channelnumber)

  def _rxcv(self, cw):
    try:
      return self._rxcvimpl(cw)
    except:
      self.logger.exception("RX[%d] Exception", self.channelnumber)

  def _Tx(self, queue, cw, txfunc, txdata, rxlen, rxfunc, rxdata):

    # TX completion object
    txcompl = None
    if txfunc:
      txcompl = TXCompletionItem(txfunc, txdata)

    # Solicited reply - reserve rxlen frames after transmission
    rxcompl = None
    if rxlen>0:
      rxcompl = RXCompletionItem(rxlen, rxfunc, rxdata)
   
    # Add to transmit queue
    queue.put(TXItem(cw, txcompl, rxcompl))

  def Start(self):
    self.modem.start()

  def Tx(self, cw, txfunc=None, txdata=None, rxlen=0, rxfunc=None, rxdata=None):
    self._Tx(self.txqueue, cw, txfunc, txdata, rxlen, rxfunc, rxdata)
  
  def TxTraf(self, cw, txfunc=None, txdata=None, rxlen=0, rxfunc=None, 
             rxdata=None):
    self._Tx(self.txtrafqueue, cw, txfunc, txdata, rxlen, rxfunc, rxdata)

if __name__=="__main__":

  def sammis(data, err, ch, cws):
    if err:
      print("*** TIMEOUT")
    else:
      print("*** SAMMIS!!!")

  def rxfunc(data, ch, o):
    pass

  def txcompl(data, ch):
    print("TX complete", data, ch)

  def gtccompl(data, ch):
    print("GTC complete.")
    return 2

  logging.basicConfig(filename="tsc-debug.log", 
                      filemode="w",
                      level=logging.DEBUG)

  ch = Channel(0x3201, 1, rxfunc, None)
  ch.Start()

  while True:
    a = input("Cmd:>")

    if a=="free":
      del(ch.modem)
    if a=="stop":
      print("Stopping...")
      ch.modem.stop()
    if a=="start":
      print("Starting...")
      ch.Start()

    if a=="g":
      ch.Tx(mpt.GTC(0, 3, 0, 1, mpt.PABXI, 0))
      ch.Tx(mpt.GTC(0, 3, 0, 1, mpt.PABXI, 0), gtccompl)

    if a=="?":
      ch.Tx(mpt.AHY(0, 3, mpt.TSCI, 1, 0, 0, 0, 0), txcompl, None)

    if a=="s":
      ch.Tx(mpt.AHYC(0, 3, mpt.TSCI, 1, 0), None, None, 1, sammis, None)

    if a=="c":
      ch.TxTraf(mpt.CLEAR(1, 0, 0, 0))
      ch.TxTraf(mpt.CLEAR(1, 0, 0, 0))

File: module/channel.c
/* SoftTSC - Software MPT1327 Trunking System Controller
* Copyright (C) 2013-2014 Paul Banks (http://paulbanks.org)
* 
* This file is part of SoftTSC
*
* SoftTSC is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* SoftTSC is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with SoftTSC.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <glib.h>
#include <string.h>
#include <math.h>

#include "channel.h"

#define SYNT 0x3B28     // 0011101100101000

const static char* const morsetable[] = {
  "A.-", "B-...", "C-.-.", "D-..", "E.", "F..-.", "G--.", "H....", "I..",
  "J.---", "K-.-", "L.-..", "M--", "N-.", "O---", "P.--.", "Q--.-", "R.-.",
  "S...", "T-", "U..-", "V...-", "W.--", "X-..-", "Y-.--", "Z--..",
  "0-----", "1.----", "2..---", "3...--", "4....-", "5.....", "6-....",
  "7--...", "8---..", "9----.", NULL
};

static void modem_tx(guint64* cw, void* userdata)
{
  MPT1327Channel* ch = userdata;
  guint64 cwtmp = ch->tx_callback(ch->userdata);
  if (cwtmp==1)
    *cw = 0xAAAAAAAAAAAA0000LL | SYNT; // Traffic channel sync word
  else if (cwtmp>1)
    *cw = mpt1327_channel_fcs_add(cwtmp);
}

static void modem_rx(guint32 bit, void* userdata)
{
  MPT1327Channel* ch = userdata;
  static int x;

  // Store new bit
  ch->rx_cw <<= 1;
  ch->rx_cw |= bit;

  // TODO: improve this. We need carrier detect for starters.
  //       probably nice to know which SYNC/SYNT was used too
  if (mpt1327_channel_fcs(ch->rx_cw>>16)==(ch->rx_cw&0xFFFF))
  {
    // Strip fcs from received data
    ch->rx_callback(ch->userdata, ch->rx_cw>>16);
  }

}

static void sound_rx(const mskmodem_sound_t* buf, 
                     gint32 samples, void* userdata)
{
  MPT1327Channel* ch = userdata; 

  if (ch->enable_bridge)
  {
    int cbremain = ch->cbsnd_size - ch->cbsnd_wr;
    if (cbremain >= samples)
      memcpy(ch->cbsnd+ch->cbsnd_wr, buf, samples*sizeof(*buf));
    else {
      memcpy(ch->cbsnd+ch->cbsnd_wr, buf, cbremain*sizeof(*buf));
      memcpy(ch->cbsnd, buf, (samples-cbremain)*sizeof(*buf));
    }
    ch->cbsnd_wr = (ch->cbsnd_wr + samples) % ch->cbsnd_size;
    ch->cbsnd_ready += samples;
  }

}

static void sound_tx(mskmodem_sound_t* buf, gint32 samples, void* userdata)
{
  MPT1327Channel* ch = userdata; 
  int i, bufavail;
  int p = 0;

  // Sound buffer (rx->tx)
  if (ch->cbsnd_ready >= samples || (ch->cbsnd_ready && ! ch->enable_bridge))
  {
    int cbremain = ch->cbsnd_size - ch->cbsnd_rd;
    p = (ch->cbsnd_ready >= samples) ? samples : ch->cbsnd_ready;

    if (cbremain >= p)
      memcpy(buf, ch->cbsnd+ch->cbsnd_rd, p*sizeof(*buf));
    else {
      memcpy(buf, ch->cbsnd+ch->cbsnd_rd, cbremain*sizeof(*buf));
      memcpy(buf+cbremain, ch->cbsnd, (p-cbremain)*sizeof(*buf));
    }

    ch->cbsnd_rd = (ch->cbsnd_rd + p) % ch->cbsnd_size;
    ch->cbsnd_ready -= p;
            
  }

  // Silence any unprocessed buffer
  while (p<samples) {
    buf[p] = 0;
    p++;
  }

  g_mutex_lock(&ch->mutex); //TODO: don't lock so long

  // Mix in tones
  if (ch->cbtone_ready) {
    MPT1327Tone* t = &ch->cbtone[ch->cbtone_rd];
    if (t->fcomp) {
      t->fcomp(t->userdata);
      t->fcomp = NULL;
    }
    for (i=0; i<samples && t->duration>0; i++) {
      mskmodem_sound_t v = buf[i] + 0.6 * MSKMODEM_SOUND_FULLSCALE *  
                           sin(2.0*G_PI*t->duration*t->freq/48000.0);
      buf[i] = tanhf(v);
      t->duration -= 1;
      if (t->duration==0) {
        ch->cbtone_rd = (ch->cbtone_rd + 1) % ch->cbtone_size;
        ch->cbtone_ready -= 1;
        t = &ch->cbtone[ch->cbtone_rd];
      }
    }
  }

  g_mutex_unlock(&ch->mutex);

}

void mpt1327_channel_queue_tone(
  MPT1327Channel* ch,
  gint16 freq, 
  gint32 duration,
  mpt1327_channel_completion_fn fcomp,
  void* userdata
)
{
  g_mutex_lock(&ch->mutex);

  // If full just bomb TODO: improve this, it will leak completions!
  if (ch->cbtone_ready >= ch->cbtone_size)
    return;

  // Put tone into buffer
  ch->cbtone[ch->cbtone_wr].freq = freq;
  ch->cbtone[ch->cbtone_wr].duration = duration;
  ch->cbtone[ch->cbtone_wr].fcomp = fcomp;
  ch->cbtone[ch->cbtone_wr].userdata = userdata;
  ch->cbtone_wr = (ch->cbtone_wr + 1) % ch->cbtone_size;
  ch->cbtone_ready += 1;

  g_mutex_unlock(&ch->mutex);

}  

void mpt1327_channel_queue_morse(
  MPT1327Channel* ch, 
  const char* str,
  mpt1327_channel_completion_fn fcomp,
  void* userdata
)
{

  const int c = 3200;

  const char* s = str;
  while (*s) {

    // Look up in table
    const char* const* p = morsetable;
    while (*p) {
      if (**p==*s) {
        const char* r = *p+1;
        while (*r) {
          if (*r=='.')
            mpt1327_channel_queue_tone(ch, 800, 1 * c, NULL, NULL);
          else
            mpt1327_channel_queue_tone(ch, 800, 3 * c, NULL, NULL);
          r++;
          // Signaling space (ITU-R M.1667-1 2009 2.2)
          mpt1327_channel_queue_tone(ch, 0, 1 * c, NULL, NULL);
        }
        break;
      }
      p++;
    }

    // Letter space is 3 dots (including signal space above)
    mpt1327_channel_queue_tone(ch, 0, 2 * c, NULL, NULL);

    // Word space is 7 dots (including letter space and signal space above)
    if (*s==' ')
      mpt1327_channel_queue_tone(ch, 0, 4 * c, NULL, NULL);

    s++;
  }
  
  // Queue completion callback (TODO: improve readability)
  if (fcomp)
    mpt1327_channel_queue_tone(ch, 0, 4 * c, fcomp, userdata);

}

void mpt1327_channel_bridge(
  MPT1327Channel* ch,
  int bridge       
)
{
  ch->enable_bridge = bridge;
}

int
mpt1327_channel_stop(
  MPT1327Channel* ch
)
{
  return mskmodem_stop(ch->modem);
}

int mpt1327_channel_start(MPT1327Channel* ch)
{
  return mskmodem_run(ch->modem);
}

guint16 mpt1327_channel_fcs(guint64 cw) {

  guint64 ck = 0;
  int parity = 0, n, b;

  // Calculate checksum
  for (n=0;n<48;n++) {
    b = cw >> (47-n) & 1;
    parity ^= b;
    if ((b ^ (ck >> 15)) & 1) 
      ck ^= 0x6815;
    ck <<= 1;
  }
  ck = (ck ^ 0x0002) & 0xFFFE;

  // Calculate parity
  for (n=0;n<16;n++)
    parity ^= ck >> (15-n) & 1;

  return ck | parity;
}

guint64 mpt1327_channel_fcs_add(guint64 cw) {
  guint16 m = mpt1327_channel_fcs(cw);
  return cw<<16 | m;
}

int 
mpt1327_channel_init( 
  MPT1327Channel** ppCh,
  const char* channelId,
  mpt1327_channel_recv_fn recvfn,
  mpt1327_channel_txcv_fn txcvfn,
  void* context
)
{
  
  MPT1327Channel* ch = g_new0(MPT1327Channel, 1);
  
  mskmodem_init(&ch->modem, channelId,
                modem_rx, modem_tx,
                sound_rx, sound_tx, ch);

  ch->userdata = context;
  ch->rx_callback = recvfn;
  ch->tx_callback = txcvfn;

  g_mutex_init(&ch->mutex);

  // Sound bridge circular buffer
  ch->cbsnd_size = 10240; //TODO: arbitrary value: calculate properly
  ch->cbsnd = g_new(mskmodem_sound_t, ch->cbsnd_size);

  // Tones queue
  ch->cbtone_size = 512;
  ch->cbtone = g_new(MPT1327Tone, ch->cbtone_size);

  *ppCh = ch;

  return 0;

}

void mpt1327_channel_free(MPT1327Channel** ppCh)
{
  if (ppCh && *ppCh)
  {
    MPT1327Channel* ch = *ppCh;
    mpt1327_channel_stop(ch);
    mskmodem_free(&ch->modem);
    g_free(ch->cbsnd);
    g_free(ch->cbtone);
    g_free(ch);
    *ppCh = NULL;
  }
}  

